### ISSUE TITLE
[16 enterprise] Field Service task: error while trying to delete generated Sale Order

### ISSUE THREAD

**Issue by razvanioan** [odoo/odoo#116222]
Impacted versions:
- 16.0

Steps to reproduce:
- cancel Sale Order generated from Field Service task
- try to delete it

Current behavior:
- raises an error

```log
Traceback (most recent call last):
  File "/data/build/odoo/odoo/http.py", line 1584, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/data/build/odoo/odoo/service/model.py", line 134, in retrying
    result = func()
  File "/data/build/odoo/odoo/http.py", line 1611, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/data/build/odoo/odoo/http.py", line 1808, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/data/build/odoo/addons/website/models/ir_http.py", line 235, in _dispatch
    response = super()._dispatch(endpoint)
  File "/data/build/odoo/odoo/addons/base/models/ir_http.py", line 149, in _dispatch
    result = endpoint(**request.params)
  File "/data/build/odoo/odoo/http.py", line 699, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/data/build/odoo/addons/web/controllers/dataset.py", line 42, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/data/build/odoo/addons/web/controllers/dataset.py", line 33, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/data/build/odoo/odoo/api.py", line 461, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/data/build/odoo/odoo/api.py", line 448, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/data/build/odoo/addons/rating/models/rating_mixin.py", line 117, in unlink
    result = super(RatingMixin, self).unlink()
  File "/data/build/odoo/addons/event_sale/models/sale_order.py", line 59, in unlink
    return super(SaleOrder, self).unlink()
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 330, in unlink
    res = super(MailThread, self).unlink()
  File "/data/build/odoo/addons/mail/models/mail_activity_mixin.py", line 246, in unlink
    result = super(MailActivityMixin, self).unlink()
  File "/data/build/odoo/odoo/models.py", line 3591, in unlink
    self.env.flush_all()
  File "/data/build/odoo/odoo/api.py", line 734, in flush_all
    self[model_name].flush_model()
  File "/data/build/odoo/odoo/models.py", line 5562, in flush_model
    self._flush(fnames)
  File "/data/build/odoo/odoo/models.py", line 5624, in _flush
    assert len(values) == len(records), \
AssertionError: Could not find all values of sale.order.line.order_partner_id to flush them
    Context: {'lang': 'en_US', 'tz': 'Asia/Calcutta', 'uid': 2, 'allowed_company_ids': [1], 'fsm_mode': True, 'active_model': 'project.task', 'active_id': 114, 'active_ids': [114], 'create': False, 'show_sale': True}
    Cache: {account.analytic.line.account_id: {563: 27}, account.analytic.line.amount: {563: -500.0}, account.analytic.line.ancestor_task_id: {563: None}, account.analytic.line.category: {563: 'other'}, account.analytic.line.code: {563: None}, account.analytic.line.company_id: {563: 1}, account.analytic.line.create_date: {563: datetime.datetime(2023, 3, 22, 14, 42, 36, 563447)}, account.analytic.line.create_uid: {563: 2}, account.analytic.line.currency_id: {563: 2}, account.analytic.line.date: {563: datetime.date(2023, 3, 22)}, account.analytic.line.department_id: {563: 3}, account.analytic.line.employee_id: {563: 1}, account.analytic.line.general_account_id: {563: None}, account.analytic.line.global_leave_id: {563: None}, account.analytic.line.helpdesk_ticket_id: {563: None}, account.analytic.line.holiday_id: {563: None}, account.analytic.line.is_so_line_edited: {563: False}, account.analytic.line.journal_id: {563: None}, account.analytic.line.manager_id: {563: None}, account.analytic.line.move_line_id: {563: None}, account.analytic.line.name: {563: 'Test'}, account.analytic.line.order_id: {563: None}, account.analytic.line.partner_id: {563: 10}, account.analytic.line.plan_id: {563: 2}, account.analytic.line.product_id: {563: None}, account.analytic.line.product_uom_id: {563: 4}, account.analytic.line.project_id: {563: 9}, account.analytic.line.ref: {563: None}, account.analytic.line.slot_id: {563: None}, account.analytic.line.so_line: {563: None}, account.analytic.line.task_id: {563: 114}, account.analytic.line.timesheet_invoice_id: {563: None}, account.analytic.line.timesheet_invoice_type: {563*: 'non_billable'}, account.analytic.line.unit_amount: {563: 5.0}, account.analytic.line.user_id: {563: 2}, account.analytic.line.validated: {563: None}, account.analytic.line.write_date: {563: datetime.datetime(2023, 3, 22, 14, 42, 37, 46483)}, account.analytic.line.write_uid: {563: 2}, pos.order.line.sale_order_line_id: {}, project.project.access_token: {9: None}, project.project.active: {9: True}, project.project.alias_id: {9: 47}, project.project.allocated_hours: {9: None}, project.project.allow_billable: {9: True}, project.project.allow_forecast: {9: False}, project.project.allow_material: {9: True}, project.project.allow_milestones: {9: False}, project.project.allow_quotations: {9: False}, project.project.allow_recurring_tasks: {9: False}, project.project.allow_subtasks: {9: False}, project.project.allow_task_dependencies: {9: False}, project.project.allow_timesheets: {9: True}, project.project.allow_worksheets: {9: True}, project.project.analytic_account_id: {9: 27}, project.project.color: {9: None}, project.project.company_id: {9: 1}, project.project.create_date: {9: datetime.datetime(2023, 3, 22, 13, 30, 56, 938488)}, project.project.create_uid: {9: 1}, project.project.date: {9: None}, project.project.date_start: {9: None}, project.project.description: {9: None}, project.project.documents_folder_id: {9: 53}, project.project.is_fsm: {9: True}, project.project.label_tasks: {9: {'en_US': 'Tasks'}}, project.project.last_update_id: {9: None}, project.project.last_update_status: {9: 'to_define'}, project.project.message_main_attachment_id: {9: None}, project.project.name: {9: {'en_US': 'Field Service'}}, project.project.partner_email: {9: None}, project.project.partner_id: {9: None}, project.project.partner_phone: {9: None}, project.project.pricing_type: {9: 'task_rate'}, project.project.privacy_visibility: {9: 'portal'}, project.project.rating_active: {9: False}, project.project.rating_request_deadline: {9: datetime.datetime(2023, 4, 21, 13, 30, 57, 921760)}, project.project.rating_status: {9: 'stage'}, project.project.rating_status_period: {9: 'monthly'}, project.project.sale_line_employee_ids: {9: ()}, project.project.sale_line_id: {9: None}, project.project.sale_order_id: {9: None}, project.project.sequence: {9: 10}, project.project.task_properties_definition: {9: None}, project.project.timesheet_product_id: {9: 211}, project.project.use_documents: {9: False}, project.project.user_id: {9: 1}, project.project.worksheet_template_id: {9: 3}, project.project.write_date: {9: datetime.datetime(2023, 3, 22, 14, 39, 50, 148206)}, project.project.write_uid: {9: 2}, project.sale.line.employee.map.project_id: {}, project.task.access_token: {114: '113b8ae3-d99f-45d1-a746-7eaeda239a00'}, project.task.active: {114: True}, project.task.allocated_hours: {114: 0.0}, project.task.allow_billable: {114: True}, project.task.analytic_account_id: {114: 27}, project.task.ancestor_id: {114: None}, project.task.color: {114: None}, project.task.comment: {114: None}, project.task.commercial_partner_id: {114: 10}, project.task.company_id: {114: 1}, project.task.create_date: {114: datetime.datetime(2023, 3, 22, 14, 42, 36, 563447)}, project.task.create_uid: {114: 2}, project.task.date_assign: {114: datetime.datetime(2023, 3, 22, 14, 42, 36)}, project.task.date_deadline: {114: None}, project.task.date_end: {114: datetime.datetime(2023, 3, 22, 14, 42, 37)}, project.task.date_last_stage_update: {114: datetime.datetime(2023, 3, 22, 14, 42, 37)}, project.task.description: {114: None}, project.task.display_project_id: {114: 9}, project.task.displayed_image_id: {114: None}, project.task.effective_hours: {114: 5.0}, project.task.email_cc: {114: None}, project.task.email_from: {114: 'info@agrolait.com'}, project.task.fsm_done: {114: True}, project.task.fsm_is_sent: {114: None}, project.task.helpdesk_ticket_id: {114: None}, project.task.is_analytic_account_id_changed: {114: None}, project.task.is_blocked: {114: False}, project.task.is_closed: {114: True}, project.task.kanban_state: {114: 'normal'}, project.task.message_main_attachment_id: {114: None}, project.task.milestone_id: {114: None}, project.task.name: {114: 'Test'}, project.task.overtime: {114: 0.0}, project.task.parent_id: {114: None}, project.task.partner_email: {114: 'info@agrolait.com'}, project.task.partner_id: {114: 10}, project.task.partner_phone: {114: '+32 10 588 558'}, project.task.planned_date_begin: {114: None}, project.task.planned_date_end: {114: None}, project.task.planned_hours: {114: 0.0}, project.task.priority: {114: '0'}, project.task.progress: {114: 0.0}, project.task.project_id: {114: 9}, project.task.rating_last_value: {114: 0.0}, project.task.recurrence_id: {114: None}, project.task.recurring_task: {114: None}, project.task.remaining_hours: {114: -5.0}, project.task.sale_line_id: {114: None}, project.task.sale_order_id: {114: None}, project.task.sequence: {114: 10}, project.task.stage_id: {114: 3}, project.task.subtask_effective_hours: {114: 0.0}, project.task.total_hours_spent: {114: 5.0}, project.task.working_days_close: {114: 0.0}, project.task.working_days_open: {114: 0.0}, project.task.working_hours_close: {114: 0.0}, project.task.working_hours_open: {114: 0.0}, project.task.worksheet_signed_by: {114: None}, project.task.worksheet_template_id: {114: 3}, project.task.write_date: {114: datetime.datetime(2023, 3, 22, 14, 42, 37, 46483)}, project.task.write_uid: {114: 2}, res.partner.activation: {10: None}, res.partner.active: {10: True}, res.partner.additional_info: {10: None}, res.partner.amazon_email: {10: None}, res.partner.assigned_partner_id: {10: 9}, res.partner.associate_member: {10: None}, res.partner.avalara_partner_code: {10: None}, res.partner.box_1099_id: {10: None}, res.partner.calendar_last_notif_ack: {10: datetime.datetime(2023, 3, 22, 13, 25, 23)}, res.partner.citizen_identification: {10: None}, res.partner.city: {10: 'Pleasant Hill'}, res.partner.city_id: {10: None}, res.partner.color: {10: 0}, res.partner.comment: {10: None}, res.partner.commercial_company_name: {10: 'Deco Addict'}, res.partner.commercial_partner_id: {10: 10}, res.partner.commission_plan_id: {10: None}, res.partner.company_id: {10: None}, res.partner.company_name: {10: None}, res.partner.company_registry: {10: None}, res.partner.contact_address_complete: {10: '77 Santa Barbara Rd, 94523 Pleasant Hill, California, United States'}, res.partner.country_id: {10: 233}, res.partner.create_date: {10: datetime.datetime(2023, 3, 22, 13, 24, 55, 686653)}, res.partner.create_uid: {10: 1}, res.partner.customer_rank: {10: 20}, res.partner.date: {10: None}, res.partner.date_localization: {10: None}, res.partner.date_partnership: {10: None}, res.partner.date_review: {10: None}, res.partner.date_review_next: {10: None}, res.partner.debit_limit: {10: None}, res.partner.display_name: {10: 'Deco Addict'}, res.partner.ebay_id: {10: None}, res.partner.email: {10: 'info@agrolait.com'}, res.partner.email_normalized: {10: 'info@agrolait.com'}, res.partner.employee: {10: None}, res.partner.followup_reminder_type: {10: 'automatic'}, res.partner.free_member: {10: None}, res.partner.function: {10: None}, res.partner.grade_id: {10: None}, res.partner.grade_sequence: {10: None}, res.partner.implemented_partner_count: {10: 1}, res.partner.industry_id: {10: None}, res.partner.invoice_warn: {10: 'no-message'}, res.partner.invoice_warn_msg: {10: None}, res.partner.is_company: {10: True}, res.partner.is_published: {10: True}, res.partner.l10n_se_check_vendor_ocr: {10: None}, res.partner.l10n_se_default_vendor_payment_ref: {10: None}, res.partner.lang: {10: 'fr_FR'}, res.partner.last_time_entries_checked: {10: None}, res.partner.membership_amount: {10: None}, res.partner.membership_cancel: {10: None}, res.partner.membership_start: {10: None}, res.partner.membership_state: {10: 'none'}, res.partner.membership_stop: {10: None}, res.partner.message_bounce: {10: 0}, res.partner.message_main_attachment_id: {10: None}, res.partner.mobile: {10: None}, res.partner.monetary: {10: None}, res.partner.name: {10: 'Deco Addict'}, res.partner.ocn_token: {10: None}, res.partner.online_partner_information: {10: None}, res.partner.parent_id: {10: None}, res.partner.partner_gid: {10: None}, res.partner.partner_latitude: {10: None}, res.partner.partner_longitude: {10: None}, res.partner.partner_share: {10: True}, res.partner.partner_weight: {10: 0}, res.partner.phone: {10: '+32 10 588 558'}, res.partner.phone_sanitized: {10: '+3210588558'}, res.partner.picking_warn: {10: 'no-message'}, res.partner.picking_warn_msg: {10: None}, res.partner.plan_to_change_bike: {10: None}, res.partner.plan_to_change_car: {10: None}, res.partner.purchase_warn: {10: 'no-message'}, res.partner.purchase_warn_msg: {10: None}, res.partner.ref: {10: None}, res.partner.sale_warn: {10: 'no-message'}, res.partner.sale_warn_msg: {10: None}, res.partner.sanitized_mobile: {10: ''}, res.partner.sanitized_phone: {10: '+3210588558'}, res.partner.seo_name: {10: None}, res.partner.signup_expiration: {10: None}, res.partner.signup_token: {10: 'Vw48DD9hJu07PT1SjQjc'}, res.partner.signup_type: {10: 'signup'}, res.partner.siret: {10: None}, res.partner.state_id: {10: 13}, res.partner.street: {10: '77 Santa Barbara Rd'}, res.partner.street2: {10: None}, res.partner.street_name: {10: '77 Santa Barbara Rd'}, res.partner.street_number: {10: ''}, res.partner.street_number2: {10: ''}, res.partner.supplier_rank: {10: 0}, res.partner.team_id: {10: None}, res.partner.title: {10: None}, res.partner.type: {10: 'contact'}, res.partner.tz: {10: None}, res.partner.user_id: {10: None}, res.partner.vat: {10: None}, res.partner.website: {10: 'http://www.deco-addict.com'}, res.partner.website_description: {10: {'en_US': '<p>\n                Deco Addict designs, develops, integrates and supports HR and Supply\n                Chain processes in order to make our customers more productive,\n                responsive and profitable.\n                </p><p>\n                Our experts invent, imagine and develop solutions which meet\n                your business requirements. They build a new technical\n                environment for your company, but they always take the already\n                installed IT software into account. That is why Idealis\n                Consulting delivers excellence in HR and SC Management.\n                </p><p>\n                Deco Addict integrates ERP for Global Companies and supports PME\n                with Open Sources software to manage their businesses. Our\n                consultants are experts in the following areas:\n                </p>\n                <ul>\n                    <li>Sales and Distribution</li>\n                    <li>Materials Management</li>\n                    <li>Inventory and Warehouse management</li>\n                    <li>Customer Relationship Management</li>\n                    <li>Personnel Administration</li>\n                    <li>Talent Management</li>\n                    <li>Reporting</li>\n                </ul>\n            '}}, res.partner.website_id: {10: None}, res.partner.website_meta_og_img: {10: None}, res.partner.website_short_description: {10: {'en_US': 'Deco Addict designs, develops, integrates and supports HR and Supply Chain processes in order to make our customers more productive, responsive and profitable.'}}, res.partner.write_date: {10: datetime.datetime(2023, 3, 22, 14, 28, 34, 661454)}, res.partner.write_uid: {10: 2}, res.partner.zip: {10: '94523'}, sale.order.line.invoice_lines: {155: ()}, sale.order.line.pos_order_line_ids: {155: ()}, sale.order.line.qty_invoiced: {155*: 0.0}, sale.order.line.temporal_type: {155: None}}

The above server error caused the following client error:
null
```

Expected behavior:
- to delete the Sale Order

Video/Screenshot link (optional):

https://user-images.githubusercontent.com/2914226/226941463-9e5bfb20-73b2-447d-9815-cfaa21d007eb.mp4

Resolution:
- if sale order item is being removed from the task, then the Sale Order can be deleted

https://user-images.githubusercontent.com/2914226/226946934-45662c10-86d9-4bc7-90c3-753c1ee28989.mp4

**Comment by AbdelAzizMohamedMousa** [comment]
The error message suggests that the problem is caused by missing values for the order_partner_id field in the sale.order.line table. The error occurred while attempting to delete a sale order that was generated from a Field Service task.

To troubleshoot the issue, you can try the following steps:

Check if there are any orphaned records in the sale.order.line table that might be causing the issue. Orphaned records are records that have a foreign key reference to another table, but the referenced record no longer exists. If there are any orphaned records, try to delete them and see if the issue is resolved.

Check if there are any custom modules that modify the sale.order.line table or the unlink method of the sale.order object. If there are, check if they are properly handling the deletion of records that have a foreign key reference to them.

Check if there are any other modules that might be modifying the sale.order.line table or the unlink method of the sale.order object. Disable these modules one by one and see if the issue is resolved.

Check if there are any database constraints that might be causing the issue. If there are, modify or remove them as necessary.

If none of the above steps resolve the issue, try to reproduce the issue on a fresh installation of Odoo. If the issue can be reproduced on a fresh installation, it might be a bug in Odoo itself. In that case, you should report the issue to the Odoo community or the Odoo development team.

**Comment by niyasraphy** [comment]
@AbdelAzizMohamedMousa detailed explanation. 
is this the responses from chatgpt :exclamation:

**Comment by pedrobaeza** [comment]
That seems so...

**Comment by AbdelAzizMohamedMousa** [comment]
if it does help why not indeed I do use gpt to help The error message indicates that the system was unable to find all the values of sale.order.line.order_partner_id required to flush them. This error is likely occurring because there are missing or inconsistent values in the sale.order.line.order_partner_id field.

To resolve this error, you may want to try the following steps:

Check the sale.order.line.order_partner_id field in the database to ensure that there are no missing or inconsistent values. You may want to compare the values in this field with those in other related fields to identify any inconsistencies.

If there are inconsistencies or missing values, you may want to update the affected records to ensure that the sale.order.line.order_partner_id field is consistent and complete.

If the issue persists after updating the records, you may want to try restarting the Odoo server to see if that resolves the issue.

If the issue still persists, you may want to review any customizations or third-party modules that could be affecting the sale.order.line.order_partner_id field. You may want to disable any customizations or modules and test the system again to see if the issue persists.

**Comment by pedrobaeza** [comment]
Well, some of the answers given by ChatGPT are at least dubious, or directly doesn't solve the issue at all, so it should be filtered and corrected instead of being copied/pasted IMO.

**Comment by mart-e** [comment]
@AbdelAzizMohamedMousa thank you for helping but we would like to keep the discussions with answers from real humans. While ChatGPT is an impressive piece of tech, it is just not accurate to solve bugs.
For instance, here we have a traceback, indicating there is most likely a bug to solve. Your answer search for workarounds, this is not what we need, we need an actual patch to fix it.
In #114212 the issue is that the en_GB.po is outdated (from 9.0) and has the old bug solved by 0c261df33d1401d236b99213d3cd7a9fa0298a78. The answer you give is incorrect, it does not need a translation, it needs a correction.

I assume you had the intention to help but as it gives incorrect answer, it does more harm than good. Please refrain from using ChatGPT to give answer. You could use to understand an issue and give you a place to start but don't just copy-paste an answer without verifying it is a correct answer, otherwise it is just noise.
